### ISSUE TITLE
Dependency patch being applied to dependent

### ISSUE THREAD

**Issue by adamjstewart** [spack/spack#8475]
In #8263, I added a patch for JasPer needed to build GDAL:
```python
depends_on('jasper@1.900.1', patches=patch('uuid.patch'), when='+jasper')
```
However, it seems like Spack is trying to apply this patch to GDAL itself:
```console
$ spack install gdal~jasper
...
==> Installing gdal
==> Using cached archive: /Users/Adam/spack/var/spack/cache/gdal/gdal-2.3.0.tar.xz
==> Staging archive: /Users/Adam/spack/var/spack/stage/gdal-2.3.0-wagxrqcruikkh4yi32oziad4ucvyj4pe/gdal-2.3.0.tar.xz
==> Created stage in /Users/Adam/spack/var/spack/stage/gdal-2.3.0-wagxrqcruikkh4yi32oziad4ucvyj4pe
The text leading up to this was:
--------------------------
|diff -Naur jasper-1.900.1/src/libjasper/jp2/jp2_cod.c jasper-1.900.1.uuid/src/libjasper/jp2/jp2_cod.c
|--- jasper-1.900.1/src/libjasper/jp2/jp2_cod.c	2007-01-19 15:43:05.000000000 -0600
|+++ jasper-1.900.1.uuid/src/libjasper/jp2/jp2_cod.c	2007-03-06 07:49:58.000000000 -0600
--------------------------
File to patch: 
Skip this patch? [y] 
5 out of 5 hunks ignored
The text leading up to this was:
--------------------------
|diff -Naur jasper-1.900.1/src/libjasper/jp2/jp2_cod.h jasper-1.900.1.uuid/src/libjasper/jp2/jp2_cod.h
|--- jasper-1.900.1/src/libjasper/jp2/jp2_cod.h	2007-01-19 15:43:05.000000000 -0600
|+++ jasper-1.900.1.uuid/src/libjasper/jp2/jp2_cod.h	2007-03-06 07:49:58.000000000 -0600
--------------------------
File to patch: 
Skip this patch? [y] 
3 out of 3 hunks ignored
The text leading up to this was:
--------------------------
|diff -Naur jasper-1.900.1/src/libjasper/jp2/jp2_enc.c jasper-1.900.1.uuid/src/libjasper/jp2/jp2_enc.c
|--- jasper-1.900.1/src/libjasper/jp2/jp2_enc.c	2007-01-19 15:43:05.000000000 -0600
|+++ jasper-1.900.1.uuid/src/libjasper/jp2/jp2_enc.c	2007-03-06 07:49:58.000000000 -0600
--------------------------
File to patch: 
Skip this patch? [y] 
5 out of 5 hunks ignored
==> Patch uuid.patch failed.
==> Error: ProcessError: Command exited with status 1:
    '/usr/bin/patch' '-s' '-p' '1' '-i' '/Users/Adam/spack/var/spack/repos/builtin/packages/gdal/uuid.patch' '-d' '.'
==> Error: [Errno 2] No such file or directory: '/Users/Adam/spack/var/spack/stage/gdal-2.3.0-wagxrqcruikkh4yi32oziad4ucvyj4pe/gdal-2.3.0/spack-build.out'
```
This patch should be applied to `jasper`, not to `gdal`. In fact, it shouldn't be applied at all since I'm explicitly disabling the `jasper` variant.

@scheibelp Can you take a look at this? I checked out the commits before and after #8272 but it didn't seem to be the cause of this bug. I swear I didn't have this problem when I wrote #8263 (because the patch wasn't being applied, to `gdal` or to `jasper`).

**Comment by neilflood** [comment]
Thank you Adam. I had also run into this Jasper patch weirdness, and was planning to investigate today. The one thing I can add is that I don't get this behaviour if I am building for Python-2 bindings, and running spack under Python-2. To build bindings for Python-3 I need to run spack under Python-3 (for reasons not yet clear to me), and under these circumstances I get the Jasper patch problem. I don't know what this means, but thought I would mention it. 

(Aside: I have only recently discovered spack, and while there are some imperfections, I have to say it is awesome!)

**Comment by scheibelp** [comment]
I'll be able to check into this more tomorrow, 

`spack install gdal~jasper` doesn't appear to try to patch `gdal` for me on 98d8bdebbe9e

```
...
==> Installing gdal
==> Using cached archive: /g/g17/scheibel/repos/destdir/spack/var/spack/cache/gdal/gdal-2.3.0.tar.xz
==> Staging archive: /g/g17/scheibel/repos/destdir/spack/var/spack/stage/gdal-2.3.0-xhv5scdslb7z5k3tedrwmr7liujas7er/gdal-2.3.0.tar.xz
==> Created stage in /g/g17/scheibel/repos/destdir/spack/var/spack/stage/gdal-2.3.0-xhv5scdslb7z5k3tedrwmr7liujas7er
==> No patches needed for gdal
```

in the meantime what does the following give you?

```
$ spack-python
>>> import spack.spec
>>> gdal = spack.spec.Spec('gdal+jasper')
>>> gdal.concretize()
>>> gdal.patches
[]
>>> print list(x.path_or_url for x in gdal['jasper'].patches)
['fix_alpha_channel_assert_fail.patch', 'uuid.patch']
>>> gdal.package.do_stage()
>>> gdal.package.do_patch() # should be "no patches"
```

**Comment by adamjstewart** [comment]
> The one thing I can add is that I don't get this behaviour if I am building for Python-2 bindings, and running spack under Python-2. 

Thanks @neilflood, you're completely right, this appears to be a Python 3 specific problem. @scheibelp can you see if you can reproduce this using Python 3?

> To build bindings for Python-3 I need to run spack under Python-3 (for reasons not yet clear to me)

Hmm, this should not be the case, you should be able to say:
```console
$ spack install gdal+python ^python@3:
```
and have things work even if you are running Spack using Python 2.

> in the meantime what does the following give you?

Using Python 3, I'm seeing:
```console
$ spack-python
Spack version 0.11.2
Python 3.6.4, Darwin x86_64
>>> import spack.spec
>>> gdal = spack.spec.Spec('gdal+jasper')
>>> gdal.concretize()
>>> gdal.patches
[<spack.patch.FilePatch object at 0x1061bd6a0>]
>>> print([x.path_or_url for x in gdal['jasper'].patches])
['fix_alpha_channel_assert_fail.patch', 'uuid.patch']
>>> gdal.package.do_stage()
==> Using cached archive: /Users/Adam/spack/var/spack/cache/gdal/gdal-2.3.0.tar.xz
==> Staging archive: /Users/Adam/spack/var/spack/stage/gdal-2.3.0-rltcov3wn32vqvggrctq22w2iibn5zdv/gdal-2.3.0.tar.xz
==> Created stage in /Users/Adam/spack/var/spack/stage/gdal-2.3.0-rltcov3wn32vqvggrctq22w2iibn5zdv
>>> gdal.package.do_patch()
==> Using cached archive: /Users/Adam/spack/var/spack/cache/gdal/gdal-2.3.0.tar.xz
==> Already staged gdal-2.3.0-rltcov3wn32vqvggrctq22w2iibn5zdv in /Users/Adam/spack/var/spack/stage/gdal-2.3.0-rltcov3wn32vqvggrctq22w2iibn5zdv
The text leading up to this was:
--------------------------
|diff -Naur jasper-1.900.1/src/libjasper/jp2/jp2_cod.c jasper-1.900.1.uuid/src/libjasper/jp2/jp2_cod.c
|--- jasper-1.900.1/src/libjasper/jp2/jp2_cod.c	2007-01-19 15:43:05.000000000 -0600
|+++ jasper-1.900.1.uuid/src/libjasper/jp2/jp2_cod.c	2007-03-06 07:49:58.000000000 -0600
--------------------------
File to patch: 
Skip this patch? [y] 
5 out of 5 hunks ignored
The text leading up to this was:
--------------------------
|diff -Naur jasper-1.900.1/src/libjasper/jp2/jp2_cod.h jasper-1.900.1.uuid/src/libjasper/jp2/jp2_cod.h
|--- jasper-1.900.1/src/libjasper/jp2/jp2_cod.h	2007-01-19 15:43:05.000000000 -0600
|+++ jasper-1.900.1.uuid/src/libjasper/jp2/jp2_cod.h	2007-03-06 07:49:58.000000000 -0600
--------------------------
File to patch: 
Skip this patch? [y] 
3 out of 3 hunks ignored
The text leading up to this was:
--------------------------
|diff -Naur jasper-1.900.1/src/libjasper/jp2/jp2_enc.c jasper-1.900.1.uuid/src/libjasper/jp2/jp2_enc.c
|--- jasper-1.900.1/src/libjasper/jp2/jp2_enc.c	2007-01-19 15:43:05.000000000 -0600
|+++ jasper-1.900.1.uuid/src/libjasper/jp2/jp2_enc.c	2007-03-06 07:49:58.000000000 -0600
--------------------------
File to patch: 
Skip this patch? [y] 
5 out of 5 hunks ignored
==> Patch uuid.patch failed.
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/Users/Adam/spack/lib/spack/spack/package.py", line 1134, in do_patch
    patch.apply(self.stage)
  File "/Users/Adam/spack/lib/spack/spack/patch.py", line 104, in apply
    "-d", self.working_dir)
  File "/Users/Adam/spack/lib/spack/spack/util/executable.py", line 183, in __call__
    proc.returncode, cmd_line)
spack.util.executable.ProcessError: Command exited with status 1:
    '/usr/bin/patch' '-s' '-p' '1' '-i' '/Users/Adam/spack/var/spack/repos/builtin/packages/gdal/uuid.patch' '-d' '.'
```

**Comment by neilflood** [comment]
@adamjstewart. re: the inability to build for Python-3 when running under Python-2, this has happened for a number of packages, not just gdal. Below is a simple example,and this is always the same error. However, now that I have pasted this traceback in I am reading it properly, and I see that there is a mixture of Python versions involved, which is very wrong. So, I suspect that I have somehow managed to create some incompatible builds, or an inconsistent environment. I will start all over again and see how I go. 

```console
$ python -V
2.7.14

$ spack install py-requests^python@3:
.......
==> Installing py-requests
Failed to import the site module
Traceback (most recent call last):
  File "/scratch/rsc7/floodn/spackInstall2/opt/spack/linux-sles12-x86_64/gcc-4.8/py-snuggs-1.4.1-hfod5mbupr6urq34aid7c2brutgycium/lib/python2.7/site-packages/site.py", line 73, in <module>
    __boot()
  File "/scratch/rsc7/floodn/spackInstall2/opt/spack/linux-sles12-x86_64/gcc-4.8/py-snuggs-1.4.1-hfod5mbupr6urq34aid7c2brutgycium/lib/python2.7/site-packages/site.py", line 26, in __boot
    import imp  # Avoid import loop in Python >= 3.3
  File "/scratch/rsc7/floodn/spackInstall2/opt/spack/linux-sles12-x86_64/gcc-6/python-3.6.5-ggx2dg2px7r6z2zgtezzjfgxxqn6snrk/lib/python3.6/imp.py", line 27, in <module>
    import tokenize
  File "/scratch/rsc7/floodn/spackInstall2/opt/spack/linux-sles12-x86_64/gcc-6/python-3.6.5-ggx2dg2px7r6z2zgtezzjfgxxqn6snrk/lib/python3.6/tokenize.py", line 33, in <module>
    import re
  File "/scratch/rsc7/floodn/spackInstall2/opt/spack/linux-sles12-x86_64/gcc-6/python-3.6.5-ggx2dg2px7r6z2zgtezzjfgxxqn6snrk/lib/python3.6/re.py", line 142, in <module>
    class RegexFlag(enum.IntFlag):
AttributeError: module 'enum' has no attribute 'IntFlag'
==> Error: ProcessError: Command exited with status 1:
    '/scratch/rsc7/floodn/spackInstall2/opt/spack/linux-sles12-x86_64/gcc-6/python-3.6.5-ggx2dg2px7r6z2zgtezzjfgxxqn6snrk/bin/python3.6' '-c' 'from distutils.sysconfig import get_config_var; print(get_config_var('"'"'prefix'"'"'))'

```

**Comment by adamjstewart** [comment]
@neilflood I wasn't able to reproduce your problem, but let's not get side-tracked. Do you want to open up a separate issue? You may also be interested in #7926 if you plan on making heavy use of Python 3.

**Comment by neilflood** [comment]
@adamjstewart. I agree, this is a side-track, and I now think almost certainly due to some screw-up on my part. If I find otherwise, I will open a separate issue. Thank you for the pointer to #7926, though.

**Comment by scheibelp** [comment]
Ok so I think I've got most of this figured out: short version is that instead of saying 

```
depends_on('jasper@1.900.1', patches=patch('uuid.patch'), when='+jasper')
```

it should be

```
depends_on('jasper@1.900.1', patches='uuid.patch', when='+jasper')
``` 

I think for former works in python2 and doesn't in python3 because of the order that directives are processed in (if the `depends_on` directives are processed first, then it works) - I haven't 100% confirmed this though. The latter appears to work in both cases. Perhaps the former ought to be supported, but I see no guarantee that using `patch()` in `depends_on` should work in either python2 or python3 (it just works by happenstance in python2).

I think it would make sense for now to update the `depends_on` directive processing to prohibit using the `patch()` function as an argument to `depends_on`. Either way, `gdal` itself will also need to change.

**Comment by adamjstewart** [comment]
In that case I think there are several packages that need to be updated. I just stole this syntax from another package.

**Comment by adamjstewart** [comment]
Haven't encountered this since 2018, let's hope it got fixed somewhere along the way 🙂
